src/librustc_middle/ty/context.rs

```diff
@@ -2589,6 +2590,22 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 }
 
+impl TyCtxtAt<'tcx> {
+    /// Constructs a `TyKind::Error` type and registers a `delay_span_bug` to ensure it gets used.
+    #[track_caller]
+    pub fn ty_error(self) -> Ty<'tcx> {
+        self.tcx.ty_error_with_message(self.span, "TyKind::Error constructed but no error reported")
+    }
+
+    /// Constructs a `TyKind::Error` type and registers a `delay_span_bug` with the given `msg to
+    /// ensure it gets used.
+    #[track_caller]
+    pub fn ty_error_with_message(self, msg: &str) -> Ty<'tcx> {
+        self.sess.delay_span_bug(self.span, msg);
+        self.mk_ty(Error(super::sty::DelaySpanBugEmitted(())))
```

Comment:
I'm worried about `DelaySpanBugEmitted` being defined in `ty::sty`, maybe it should be defined in `ty::context` so that only `ty::context` can create it.

Reply:
I think the reason I put it where I did is that it needs to be in the type definition of `TyKind`...

Reply:
@mark-i-m that's where it needs to be *used*, but it's a public type, you can use it from anywhere.

Reply:
@eddyb Ah, I think I see what you mean. Opened #74399